src/python/gudhi/sklearn/rips_persistence.py

```diff
@@ -0,0 +1,148 @@
+        return self
+
+    def __transform(self, inputs):
+        # persistence_dim_max is False by default, but True when expand_extra_dimension is False
+        persistence_dim_max = self.expand_extra_dimension
+        max_dimension = max(self.dim_list_)
+        if self.expand_extra_dimension:
+            max_dimension += 1
+
+        # nb_collapse "automatic" case management
+        if self.nb_collapse < 0:
+            nb_collapse = 1
+        else:
+            nb_collapse = self.nb_collapse
+        
+        if self.input_type == 'point cloud':
+            rips = RipsComplex(points=inputs, max_edge_length = self.max_edge_length)
+        elif self.input_type == 'lower distance matrix':
+            rips = RipsComplex(distance_matrix=inputs, max_edge_length = self.max_edge_length)
+        else:
+            raise ValueError("Only 'point cloud' and  'lower distance matrix' are valid input_type")
+        
+        if max_dimension > 1:
+            stree = rips.create_simplex_tree(max_dimension=1)
+            stree.collapse_edges(nb_iterations = nb_collapse)
+            stree.expansion(max_dimension)
+        else:
+            stree = rips.create_simplex_tree(max_dimension=max_dimension)
+
+        # Specific case where, despite expansion(max_dimension), stree has a lower dimension
+        if max_dimension > stree.dimension():
+            persistence_dim_max = True
+
+        stree.compute_persistence(
+            homology_coeff_field=self.homology_coeff_field,
+            min_persistence=self.min_persistence,
+            persistence_dim_max=persistence_dim_max
+        )
+
+        return [
+            stree.persistence_intervals_in_dimension(dim) for dim in self.dim_list_
+        ]
+
+    def transform(self, X, Y=None):
+        """Compute all the Vietoris-Rips complexes and their associated persistence diagrams.
+
+        :param X: list of point clouds as Euclidean coordinates or distance matrices.
+        :type X: list of list of float OR list of numpy.ndarray
+
+        :return: Persistence diagrams in the format:
+
+              - If `homology_dimensions` was set to `n`: `[array( Hn(X[0]) ), array( Hn(X[1]) ), ...]` 
+              - If `homology_dimensions` was set to `[i, j]`: `[[array( Hi(X[0]) ), array( Hj(X[0]) )], [array( Hi(X[1]) ), array( Hj(X[1]) )], ...]`
+        :rtype: list of numpy ndarray of shape (,2) or list of list of numpy ndarray of shape (,2)
+        """
+        # Depends on homology_dimensions is an integer or a list of integer (else case)
+        if isinstance(self.homology_dimensions, int):
+            unwrap = True
+            self.dim_list_ = [ self.homology_dimensions ]
+        else:
+            unwrap = False
+            self.dim_list_ = self.homology_dimensions
+
+        # threads is preferred as Rips construction and persistence computation releases the GIL
+        res = Parallel(n_jobs=self.n_jobs, prefer="threads")(delayed(self.__transform)(inputs) for inputs in X)
```

Comment:
I think a `parallel_config` can override the `prefer` to a backend using processes, which will imply calling the `__get_state__` and `__set_state__` of the simplextree, which can have a large performance impact. What do you think about `backend=threading`  instead ?

Reply:
Also, I usually like to put a `tqdm` wrapper, when a `progress` flag is set to true in the class. It works decently well with joblib, and gives an estimation of the time to finish, which can be quite useful. But that implies adding a `tqdm` dependency.

Reply:
> I think a `parallel_config` can override the `prefer` to a backend using processes, which will imply calling the `__get_state__` and `__set_state__` of the simplextree, which can have a large performance impact.

Neither the input nor the output is a SimplexTree here, and I don't think intermediate objects are serialized / copied.

> What do you think about `backend=threading` instead ?

Since we control the function that is called, we could do that, yes.

Reply:
> Neither the input nor the output is a SimplexTree here, and I don't think intermediate objects are serialized / copied.
Yes, I read this code too quickly, it's not an issue indeed

Reply:
> Also, I usually like to put a `tqdm` wrapper, when a `progress` flag is set to true in the class. It works decently well with joblib, and gives an estimation of the time to finish, which can be quite useful. But that implies adding a `tqdm` dependency. What do you think about it ?

Since we have this same pattern with joblib.Parallel in many places, I think we should consider this separately from this PR, for all of them at the same time. Could you open an issue showing what your suggestion looks like? Searching a bit, I find https://github.com/joblib/joblib/issues/972, https://stackoverflow.com/questions/37804279/how-can-we-use-tqdm-in-a-parallel-execution-with-joblib, https://stackoverflow.com/q/24983493/1918193, or even just `Parallel(...,verbose=7)`, but nothing seems perfect.

Reply:
> I think a `parallel_config` can override the `prefer` to a backend using processes, which will imply calling the `__get_state__` and `__set_state__` of the simplextree, which can have a large performance impact. What do you think about `backend=threading` instead ?

From a benchmark I performed, it seems to be the same timings. I opened an [issue on joblib](https://github.com/joblib/joblib/issues/1511) to ask for it. We will do it on a separated issue on gudhi if we should use `backend="threading"` as @mglisse was explaining.

Reply:
> Since we have this same pattern with joblib.Parallel in many places, I think we should consider this separately from this PR, for all of them at the same time. Could you open an issue showing what your suggestion looks like? Searching a bit, I find [joblib/joblib#972](https://github.com/joblib/joblib/issues/972), https://stackoverflow.com/questions/37804279/how-can-we-use-tqdm-in-a-parallel-execution-with-joblib, https://stackoverflow.com/q/24983493/1918193, or even just `Parallel(...,verbose=7)`, but nothing seems perfect.

Sorry, I forgot to answer this. The code I was thinking about is something like this:
```python
class RipsPersistence:
  def __init__(..., progress=False):
    ...
    self.progress=progress
  def _compute_rips(self,pointcloud|distance_matrix):
    # code to compute rips for this point cloud
    ...
  def _collapse_edges(self,simplextree):
    # code for collapsing 
    ...
  def _some_other_transform(self,stuff):
    # some code
    ...
  def transform(self,point_clouds):
    simplextrees = Parallel(...)(delayed(self._compute_rips)(point_cloud) 
                 for point_cloud in tqdm(point_clouds, disable = not self.progress, desc="Computing Rips Complexes")
    simplextrees = Parallel(...)(delayed(self._compute_rips)(simplextree) 
                 for simplextree in tqdm(simplextrees, disable = not self.progress, desc="Collapsing edges")
    persistence_diagrams = Parallel(...)(delayed(self._some_other_transform)(simplextree) 
                 for simplextree in tqdm(simplextrees, disable = not self.progress, desc="Computing diagrams")
    return persistence_diagrams
```
I don't know if cutting the log into these steps is good or not (or should depend on a parameter) but it can be useful to know if one of these steps is the bottleneck, and change parameters accordingly. We can also consider adding a verbose level to this, or turn these 3 tqdms into 1 or ...